kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs

```diff
@@ -0,0 +1,301 @@
+            ConstantCell::create(F::zero()),
+        ],
+    ]
+}
+
+pub fn compute_bound(x: &BigUint, neg_foreign_field_modulus: &BigUint) -> BigUint {
+    let x_bound = x + neg_foreign_field_modulus;
+    assert!(x_bound < BigUint::binary_modulus());
+    x_bound
+}
+
+fn compute_witness_variables<F: PrimeField>(
+    products: &[BigUint; 3],
+    remainder: &[BigUint; 3],
+) -> [F; 7] {
+    // Numerically this function must work on BigUints or there is something
+    // wrong with our approach.  Specifically, BigUint will throw and exception
+    // if a subtraction would underflow.
+    //
+    // By working in BigUint for this part, we implicitly check our invariant
+    // that subtracting the remainder never underflows.
+    //
+    // See the foreign field multiplication RFC for more details.
+    auto_clone_array!(products);
+    auto_clone_array!(remainder);
+
+    // C1-C2: Compute components of product1
+    let (product1_hi, product1_lo) = products(1).div_rem(&BigUint::two_to_limb());
+    let (product1_hi_1, product1_hi_0) = product1_hi.div_rem(&BigUint::two_to_limb());
+
+    // C3-C5: Compute v0 = the top 2 bits of (p0 + 2^L * p10 - r0 - 2^L * r1) / 2^2L
+    //   N.b. To avoid an underflow error, the equation must sum the intermediate
+    //        product terms before subtracting limbs of the remainder.
+    let (carry0, _) = (products(0) + BigUint::two_to_limb() * product1_lo.clone()
+        - remainder(0)
+        - BigUint::two_to_limb() * remainder(1))
+    .div_rem(&BigUint::two_to_2limb());
+
+    // C6-C9: Compute v1 = the top L + 3 bits (p2 + p11 + v0 - r2) / 2^L
+    //   N.b. Same as above, to avoid an underflow error, the equation must
+    //        sum the intermediate product terms before subtracting the remainder.
+    let (carry1, _) = (products(2) + product1_hi + carry0.clone() - remainder(2))
+        .div_rem(&BigUint::two_to_limb());
+    // Compute v10 and v11
+    let (carry1_hi, carry1_lo) = carry1.div_rem(&BigUint::two_to_limb());
+    // Compute the scaled_carry1_hi
+    let scaled_carry1_hi = carry1_hi.clone() * BigUint::from(512u32); // carr1_hi * 2^9
+
+    // C10: witness data a, b, q, and r already present
+
+    [
+        product1_lo,
+        product1_hi_0,
+        product1_hi_1,
+        carry0,
+        carry1_lo,
+        carry1_hi,
+        scaled_carry1_hi,
+    ]
+    .to_fields()
+}
+
+fn compute_bound_witness_variables<F: PrimeField>(
+    sums: &[BigUint; 2],  // [sum01, sum2]
+    bound: &[BigUint; 2], // [bound01, bound2]
+) -> [F; 2] {
+    auto_clone_array!(sums);
+    auto_clone_array!(bound);
+
+    // C11: witness data is created by externally by called and multi-range-check gate
+
+    // C12-C13: Compute q'_carry01 = (s01 - q'01)/2^2L
+    let (quotient_bound_carry01, _) = (sums(0) - bound(0)).div_rem(&BigUint::two_to_2limb());
+
+    // C14-C15: Compute q'_carry2 = (s2 + q'_carry01 - q'2)/2^L
+    let (quotient_bound_carry2, _) =
```

Comment:
this is what we think could be removed right? perhaps we want an assert equal to be zero, just as a sanity check

Reply:
Right